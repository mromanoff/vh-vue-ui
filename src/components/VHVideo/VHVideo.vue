<template>
  <div class="VHVideo">
    <div class="VHVideo-player">
      <div
        v-if="!loadPlayer"
        class="VHVideo-preview"
        @click="handlePreviewClick"
      >
        <div class="VHVideo-previewImageContainer">
          <img
            class="VHVideo-previewImage"
            :src="previewImgSrc"
          >
        </div>
        <div class="VHVideo-previewPlayButtonContainer">
          <vh-icon
            class="VHVideo-previewPlayButton"
            color="white"
            name="playCircle"
          />
        </div>
      </div>
      <no-ssr v-else>
        <youtube
          ref="youtube"
          :video-id="videoId"
          :player-vars="{ autoplay: 1 }"
          v-on="$listeners"
          @ready="handleForcePlay"
          @playing="handlePlaying"
          @paused="handlePaused"
        />
      </no-ssr>
    </div>
  </div>
</template>

<script>
export default {
  name: 'VhVideoPreview',
  props: {
    /**
     * VHVideo will force play
     */
    forcePlay: {
      type: Boolean,
      default: false,
    },
    /**
     * VHVideoReview will choose quality of preview image
     */
    quality: {
      type: String,
      default: 'hqdefault',
      validator: value => {
        return value.match(/(default|sddefault|mqdefault|hqdefault|maxresdefault)/);
      },
    },
    /**
     * Youtube player will not load until clicked
     */
    lazy: {
      type: Boolean,
      default: false,
    },
    /**
     * Takes a YouTube video source for preview
     */
    videoId: {
      type: String,
      default: '',
    },
  },

  data() {
    return {
      loadPlayer: !this.lazy,
      isPlaying: false,
      isPaused: false,
    };
  },

  computed: {
    youtubePlayer() {
      return this.$refs.youtube ? this.$refs.youtube.player : null;
    },
    previewImgSrc() {
      return `https://i.ytimg.com/vi/${this.videoId}/${this.quality}.jpg`;
    },
  },
  watch: {
    loadPlayer(val) {
      if (val) this.$emit('load-player');
    },
  },
  methods: {
    handleForcePlay() {
      if (this.forcePlay) {
        this.play();
      }
    },
    handlePreviewClick() {
      this.loadPlayer = true;
    },
    handlePlaying() {
      this.isPlaying = true;
      this.isPaused = false;
    },
    handlePaused() {
      this.isPaused = true;
      this.isPlaying = false;
    },
    play() {
      this.youtubePlayer.playVideo();
    },
    pause() {
      this.youtubePlayer.pauseVideo();
    },
  },
};
</script>

<style scoped>
.VHVideo {
  /* BLOCK */
  background-color: #000;
}

.VHVideo-player {
  overflow: hidden;
  padding-bottom: 56.25%;
  position: relative;
  height: 0;
}

.VHVideo-preview {
  position: relative;
  cursor: pointer;
  background-color: #000;
  width: 100%;
  height: 100%;
}

.VHVideo-player >>> iframe,
.VHVideo-player .VHVideo-preview {
  left: 0;
  top: 0;
  height: 100%;
  width: 100%;
  position: absolute;
}

.VHVideo-previewImageContainer {
  position: relative;
  height: 100%;
  overflow: hidden;
}

.VHVideo-previewImage {
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
  width: 100%;
  height: auto;
}

.VHVideo-previewPlayButtonContainer {
  cursor: pointer;
  position: absolute;
  margin-left: auto;
  margin-right: auto;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
}

.VHVideo-previewPlayButton {
  width: 48px;
  height: 48px;
  opacity: 0.75;
}
</style>
