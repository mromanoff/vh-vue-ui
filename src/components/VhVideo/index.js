import VhVideo from './VhVideo';

export default VhVideo;
