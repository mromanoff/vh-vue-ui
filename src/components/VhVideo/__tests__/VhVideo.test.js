import { shallowMount } from '@vue/test-utils';
import VhVideo from '../VhVideo.vue';

describe('VhVideo.vue', () => {
  let wrapper;

  it.skip('renders default VhVideo', () => {
    wrapper = shallowMount(VhVideo, {
      stubs: ['vh-icon', 'no-ssr', 'youtube'],
    });
    expect(wrapper.props().quality).toBe('hqdefault');
    expect(wrapper.props().id).toBe('');
    expect(wrapper).toMatchSnapshot();
  });

  it.skip('renders modified VhVideo', () => {
    wrapper = shallowMount(VhVideo, {
      propsData: {
        id: 'Test ID',
        quality: 'mqdefault',
      },
      stubs: ['vh-icon', 'no-ssr', 'youtube'],
    });
    expect(wrapper.props().quality).toBe('mqdefault');
    expect(wrapper.props().id).toBe('Test ID');
    expect(wrapper).toMatchSnapshot();
  });
});
