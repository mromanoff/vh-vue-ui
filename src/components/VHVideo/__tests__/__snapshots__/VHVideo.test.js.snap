// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VHVideo.vue renders default VHVideo 1`] = `
<div class="VHVideo">
  <div class="VHVideo-player">
    <div class="VHVideo-preview">
      <div class="VHVideo-previewImageContainer"><img src="https://i.ytimg.com/vi//hqdefault.jpg" class="VHVideo-previewImage"></div>
      <div class="VHVideo-previewPlayButtonContainer">
        <vh-icon-stub color="white" name="playCircle" class="VHVideo-previewPlayButton"></vh-icon-stub>
      </div>
    </div>
  </div>
</div>
`;

exports[`VHVideo.vue renders modified VHVideo 1`] = `
<div class="VHVideo">
  <div class="VHVideo-player">
    <div class="VHVideo-preview">
      <div class="VHVideo-previewImageContainer"><img src="https://i.ytimg.com/vi/Test ID/mqdefault.jpg" class="VHVideo-previewImage"></div>
      <div class="VHVideo-previewPlayButtonContainer">
        <vh-icon-stub color="white" name="playCircle" class="VHVideo-previewPlayButton"></vh-icon-stub>
      </div>
    </div>
  </div>
</div>
`;
